src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+    ;; The namespace must be launched
+    (asserts!
+      (is-some (get launched-at namespace-props))
+      (err ERR_NAMESPACE_NOT_LAUNCHED))
+    ;; The sender must match the name's current owner
+    (asserts!
+      (is-eq owner contract-caller)
+      (err ERR_NAME_OPERATION_UNAUTHORIZED))
+    ;; The name must not be expired
+    (asserts!
+      (is-eq (unwrap! (is-name-lease-expired? namespace name) (err ERR_PANIC)) false)
+      (err ERR_NAME_EXPIRED))
+    ;; The name must not be in the renewal grace period
+    (asserts!
+      (is-eq (unwrap! (is-name-in-grace-period? namespace name) (err ERR_PANIC)) false)
+      (err ERR_NAME_GRACE_PERIOD))
+    ;; The name must not be revoked
+    (asserts!
+      (is-none (get revoked-at name-props))
+      (err ERR_NAME_REVOKED))
+    ;; Clear the zonefile
+    (update-zonefile-and-props
+        namespace 
+        name  
+        (get registered-at name-props)
+        (get imported-at name-props)
+        (some block-height)
+        0x)
+    (ok true)))
+
+;; NAME_RENEWAL
+;; Depending in the namespace rules, a name can expire. For example, names in the .id namespace expire after 2 years. 
+;; You need to send a NAME_RENEWAL every so often to keep your name.
+;; You will pay the registration cost of your name to the namespace's designated burn address when you renew it.
+;; When a name expires, it enters a month-long "grace period" (5000 blocks). 
+;; It will stop resolving in the grace period, and all of the above operations will cease to be honored by the BNS consensus rules.
+;; You may, however, send a NAME_RENEWAL during this grace period to preserve your name.
+;; If your name is in a namespace where names do not expire, then you never need to use this transaction.
+(define-public (name-renewal (namespace (buff 19))
+                             (name (buff 16))
+                             (stx-to-burn uint)
+                             (new-owner (optional principal))
+                             (zonefile-hash (optional (buff 20))))
+  (let (
+    (namespace-props (unwrap!
+      (map-get? namespaces ((namespace namespace)))
+      (err ERR_NAMESPACE_NOT_FOUND)))
+    (owner (unwrap!
+      (nft-get-owner? names (tuple (name name) (namespace namespace)))
+      (err ERR_NAME_NOT_FOUND))) ;; The name must exist
+    (name-props (unwrap!
+      (map-get? name-properties ((name name) (namespace namespace)))
+      (err ERR_NAME_NOT_FOUND)))) ;; The name must exist
+    ;; The namespace must be launched
+    (asserts!
+      (is-some (get launched-at namespace-props))
+      (err ERR_NAMESPACE_NOT_LAUNCHED))
+    ;; The namespace should require renewals
+    (asserts!
+      (> (get renewal-rule namespace-props) u0)
+      (err ERR_NAME_OPERATION_UNAUTHORIZED))
+    ;; The sender must match the name's current owner
+    (asserts!
+      (is-eq owner contract-caller)
+      (err ERR_NAME_OPERATION_UNAUTHORIZED))
+    ;; The name must have been registered (vs imported)
```

Comment:
This check will need to be repaired.  A name that was imported can be renewed by its owner if the namespace is launched.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/58c72297359afda14cac67254ae854c93387a583